app/services/merge_duplicate_patients.rb

```diff
@@ -0,0 +1,194 @@
+class MergeDuplicatePatients
+  def initialize(patients)
+    @patients = patients.sort_by(&:recorded_at)
+  end
+
+  def earliest_patient
+    @patients.first
+  end
+
+  def latest_patient
+    @patients.last
+  end
+
+  def merge
+    new_patient = ActiveRecord::Base.transaction do
+      new_patient = create_patient
+      create_prescription_drugs(new_patient)
+      create_medical_history(new_patient)
+      create_phone_numbers(new_patient)
+      create_patient_business_identifiers(new_patient)
+      create_encounters_and_observables(new_patient)
+      create_appointments(new_patient)
+      create_teleconsultations(new_patient)
+      mark_as_merged(new_patient)
+      new_patient.reload
+    end
+
+    Rails.logger.info "Merged patients #{@patients.pluck(:id)} into patient #{new_patient.id}"
+    new_patient
+  end
+
+  def mark_as_merged(new_patient)
+    @patients.map { |patient| patient.update!(merged_into: new_patient.id) }
+    @patients.map(&:discard_data)
+  end
+
+  def create_patient
+    attributes = {
+      id: SecureRandom.uuid,
+      full_name: latest_patient.full_name,
+      gender: latest_patient.gender,
+      reminder_consent: latest_patient.reminder_consent,
+      recorded_at: earliest_patient.recorded_at,
+      registration_facility: earliest_patient.registration_facility,
+      registration_user: earliest_patient.registration_user,
+      device_created_at: earliest_patient.device_created_at,
+      device_updated_at: earliest_patient.device_updated_at,
+      assigned_facility: latest_patient.assigned_facility,
+      address: create_address
+    }.merge(age_and_dob)
```

Comment:
If you're feeling fancy
```ruby
{
  # ...
  address: create_address,
  **age_and_dob
}
```

I like how keeping it inside the hash makes it clear that we're just setting some more attributes.

Reply:
Heh, @prabhanshuguptagit mentioned this, and I looked up the double splat at the time as well. I felt like keeping merge there was easier to read 🤷‍♀️  Not hardcore rubyist enough I guess 😁